### ISSUE TITLE
LSF Adapter: Address performance issues

### ISSUE THREAD

**Issue by ericfranz** [OSC/ood_core#40]
Need to do a little performance analysis. A better algorithm might do. However, there is a lot of parsing and string manipulation going on. Using the C library might make it faster (via Fiddle http://ruby-doc.org/stdlib-2.2.0/libdoc/fiddle/rdoc/Fiddle.html).

It was observed when testing that 30 jobs can take 1-2 seconds and 4000 jobs could take 28 seconds. This is way too long.

Many Fiddle tutorials online. One example: http://blog.honeybadger.io/use-any-c-library-from-ruby-via-fiddle-the-ruby-standard-librarys-best-kept-secret/

See http://publibfp.dhe.ibm.com/epubs/pdf/c2753121.pdf and lsb_openjobinfo() and lsb_readjobinfo() and jobInfoEnt structure. Also, here is some example code that uses the C library: https://github.com/PlatformLSF/lsf-drmaa/blob/025f9c49af48e410dc0ab0b9c611c42935dc09eb/lsf_drmaa/job.c

**Comment by ericfranz** [comment]
1. `jobInfoEnt` structure appears to be identical between 8.3 and 9.1
2. Fiddle does not provide a `time_t` value. Try using Fiddle::TYPE_LONG in its place. Or you could try this in Fiddle:

```ruby
typealias "time_t", "long"
```

**Comment by nickjer** [comment]
I feel if you go this route, it may be best to break this off into a separate gem much like pbs-ruby.

**Comment by ericfranz** [comment]
I don't think we will realize any benefits by breaking this off into a separate gem.

**Comment by nickjer** [comment]
There are a few benefits to a full LSF library being in a separate gem.

- Improvements can be added to it without directly affecting the overall `ood_core` gem. So a patch to the LSF gem won't affect every OOD installation. We would only need to tell the LSF sites to update the LSF gem.

- You can also add support for other LSF features such as getting the status of a server, queue, or node that isn't directly related to what `ood_core`'s current capabilities are, similar to what pbs-ruby currently provides.

- If the LSF gem does require dependencies in the future outside of the standard library it won't affect the dependencies required by `ood_core`. And centers that don't use LSF won't need to install those dependencies.

**Comment by ericfranz** [comment]
We can add LSF specific features to the batch object and just document it. The other two benefits won't be realized by our project. And in fact, if there ever would be a time where it would be really beneficial to separate out two gems, that's the time to do it. Anyways, if we want to have any more discussion about the gem architecture let's do that in another issue or offline.

**Comment by brianmcmichael** [comment]
PBS-Ruby is the one Gem that's been relatively stable around here and it's been used in all of our App Kit instances. Taking the time to prepare a separate LSF Gem would reduce the overhead of having to recreate the LSF code in any future iteration. Additionally, it would give sites the ability to use the adapter directly in their apps.

I'd say that a tangental benefit would be that it would give OSC the authority that goes with having created these open-source adapters, but since we're not an LSF site I guess our ability to adequately maintain the gems is questionable.

**Comment by moffatsadeghi** [comment]
GitHub Review Meeting Summary from April: 
The team reviewed performance concerns with the LSF adapter.
Details
    johrstrom: Noted the issue discusses LSF being slow at job querying and suggests C optimizations
    Emily: Asked if any similar issues have been reported since 2017
    johrstrom: Found one discourse topic where optimization was needed
    bsingleton: Questioned whether performance issues are on LSF's end rather than the adapter, expressing uncertainty about what C optimizations could be made when just submitting to CLI
    johrstrom: Mentioned the possibility of using shared object libraries with C bindings in Ruby
    bsingleton: Suggested keeping the issue open as a seed of an idea for someone with LSF access to potentially address
    Emily: Proposed documenting it as a known issue in documentation
    bsingleton: Disagreed, stating it's unusual to categorize potential optimizations as known issues when they can't be tested
    johrstrom: Showed existing documentation about LSF info memory caching and refresh rates
Conclusion
    Issue #40 kept open but not scheduled, awaiting community feedback or contributions from LSF users
